### ISSUE TITLE
static like noise, goes away on reload

### ISSUE THREAD

**Issue by BellesoftConsulting** [JoJoBond/3LAS#4]
Hi, thanks again for this great work!
Sometimes a client would be listening to an mp3 stream over WS perfectly fine (nice audio quality) and all of a sudden  it starts playing with some background "static" and quality seems to drop. The source is fine, as I am listening to it on another client on a nother machine, where everything is fine.  If I simply reload the player on the browser where I hear the bad quality.static, then it all goes back to nice quality. 

This happens also when for example in Safari when we call "FormatReader.PurgeData();"  after that call, static is heard on the audio playback. I can replicate this by simply adding a click handler to an HTML button on the playback page that calls "FormatReader.PurgeData();" from the click handler..

**Comment by JoJoBond** [comment]
I'll have to look into this to find out what causes it.
For Apple devices I have to over-provision the decoder, because else it would fail to decode due to the mp3 byte-reservoir. 
You could try to play with the "WindowSize" and the "UseFrames" parameters in the FormatReader. 
The WindowsSize determines the number of frames that are decoded in one go. The UseFrames parameter determines of many of the decoded frames are actually used. I'll have a look and try to find out if I can narrow down the issue, If you have a good method to reproduce it please let me know.

**Comment by BellesoftConsulting** [comment]
Thanks for that info! I changed those params to 64 and 32 respectively (I am using a 64Kbs stream) and it sounds great now on Safari, FF, and Chrome! thanks!

**Comment by JoJoBond** [comment]
Please note that increasing the WindowsSize will increase the CPU load on the client.
Increasing the UseFrames will increase the latency on the client side. You should try to keep both parameters as low as possible while ensuring flawless playback.

**Comment by ServeurpersoCom** [comment]
Same problem here. I use only the PCM player and after a short time the audio turn to a bad noise.

**Comment by JoJoBond** [comment]
What OS/browser are you using?
Try increasing the client side "BatchLength" constant inside the wav.js.
Changing the "ExtraEdgeLength" might help aswell.

**Comment by ServeurpersoCom** [comment]
In fact this problem is my fault. In my implementation I must be able to cut/change the stream when I change the robot and that's when it go to "crashes". Do you have the faster and cleanest way to reset the PCMAudioPlayer to put inside my stream change event ?

**Comment by JoJoBond** [comment]
I can't quite follow you. How do you switch from one stream to the other?
I haven't implemented any method that reserts the PCMAudioPlayer to start a new stream. For now you'll have to reset some internal variables yourself.
To reset the player before calling PushBuffer the first time with the new stream you need to set the "_NextTime" member of your PCMAudioPlayer instance to 0.0.
Any bath  that is still scheduled for playback before the reset will still be played, so you might get a brief period of overlap.

**Comment by ServeurpersoCom** [comment]
This is my entire audio code. The socket.io "clientsrobotaudio" send raw PCM signed 16 bits data and I use your AudioPlayer.PushBuffer(audioBuffer); to play them. When the PCM stream is stopped and started again on server side your player go to a "click click" sound. Same problem when there is a bit of javascript CPU overload.
Now I test the _NextTime = 0.0 trick. *I will add a setter method inside your PCM player code*
Next I need to add your MP3 format player inside my project:D Thanks a lot for your work

```
   var audioCtx = new AudioContext();
   var AudioPlayer = new PCMAudioPlayer();
 
   AudioPlayer.UnderrunCallback = function() {
    trace("Client", "Audio underrun", "traces");
   }
 
   SOCKETVIDEOUSERS.on("clientsrobotaudio", function(data) {
    var array = new Int16Array(data);
    var audioBuffer = audioCtx.createBuffer(1, array.length, 22050);
    var buffering = audioBuffer.getChannelData(0);
 
    for(var i = 0; i < array.length; i++)
     buffering[i] = array[i] / 32768;
 
    AudioPlayer.PushBuffer(audioBuffer);
   });
```

**Comment by JoJoBond** [comment]
Ok, I would suggest you to use the CreateBuffer method of the PCMAudioPlayer instead of creating a second AudioContext and calling createBuffer. CreateBuffer wraps createBuffer but uses the same AudioContext as the PCMAudioPlayer.

**Comment by JoJoBond** [comment]
Also note that converting int to float in JS will be very inefficient. I would suggest you to use the wav format reader. This will do the convertion natively and will perform way better. For this you will have to send a WAV header to each newly connected client.

**Comment by ServeurpersoCom** [comment]
I added this inside your PCM player and it work for me when the player go to tick tick trash, a call to ResetPlayer fix it ! :

There is a way to make a better servoloop to autotune this value ? I still have to study your code:)
I think it's easy to add a P/PI/PD or PID loop to tune the NextTime. I have the impression that you only correct in one direction and that the system is not able to recover. This is why the player go to click click and need to be resetted.

```
PCMAudioPlayer.prototype.ResetPlayer = function () {
    this._NextTime = 0.0;
};
```
Your advice is very important to me and I will test this weekend. I must reuse the AudioContext and test your WAV or better, the MP3 Player. I must use your framework correctly.
I written this first test of your audio code on my smartphone and it's a complete success for my low latency project !

**Comment by JoJoBond** [comment]
If you have laggy network or JS hickups you can try to increase the initial buffer length "_StartOffset"  of the PCMAudioPlayer. This will give you more wiggleroom when JS timing is tight.

**Comment by ServeurpersoCom** [comment]
The system must use a servoloop and must recover itself (after some bad-playing time) from laggy network or js hickups.
I'm french sorry for my bad english.
It's better to recover from a very low latency with small buffer than workaround with large buffers.

**Comment by ServeurpersoCom** [comment]
Now that I have had a first success I will learn your code and can contribute to its improvement. I can do low level code and algorithm. For Now  I'm only focused on getting a clean PCM before testing the js MP3 codec.
I already write few NAL Units to all new users to unlock Broadway H.264 Player, it's possible to do the same for a WAV Player:) But a RAW player is better because the WAV header is useless for PCM streaming.

**Comment by ServeurpersoCom** [comment]
Why you comment on line 170. This is why my underrun callback never run:D

```
            // Check if we ran out of time
            if (this._NextTime <= this._SoundContext.currentTime) {
                if (this._NextTime + AudioBuffer.duration < this._SoundContext.currentTime) {
                    //this._NextTime += AudioBuffer.duration * 1.01;
                    //return;
                }
                // In that case reschedule the playback to [StartOffset]/2.0 s from now
                //this._NextTime = this._SoundContext.currentTime;// + StartOffset / 2.0;
                //if (typeof this._UnderrunCallback === 'function')
                //	this._UnderrunCallback();
           }

```

**Comment by ServeurpersoCom** [comment]
Okay I enabled _VariSpeed and the UnderrunCallback() (removed the _ because it's public)
I get a lot of underrun calls when the bug is here.
When I reset _NextTime to 0 the bug and the under-run disappear.

**Comment by ServeurpersoCom** [comment]
A crap workaround is an automatic reset when there is underrun
Now I need to make a better automatic recover inside audio_player.js

```
   audioPlayer.UnderrunCallback = function() {
    if($.now() - lastUnderrun > 2000) {
     trace("Client", "Audio underrun", "traces");
     setTimeout(function() {
      trace("Client", "Audio reset", "traces");
      audioPlayer.ResetPlayer();
     }, 1000);
     lastUnderrun = $.now();
    }
   }
```

**Comment by JoJoBond** [comment]
Detecting an underrun and recovering from it is quite difficult. The problem I had, was that when I got small network lags, it would reschedule my playbacks even though I was just slightly behind. I think a multi-stage approach is better, where depending on the severity of the underrun different recovery methods are used.

**Comment by JoJoBond** [comment]
Varispeed was an idea to fix Quartz to Quartz drift. It's only relevant if you wish to do very long duration transmission (multiple days) without worrying about the buffer under- or overruning.

**Comment by ServeurpersoCom** [comment]
Very interesting I just learn the AudioContext API and I understand it.
You use the decodeAudioData() to convert int16 from PCM to the crap float32 sample used by the API (they use float like for canva !).

But for now, before WAV or MP3, I need a perfect PCM player with fast recovery, I think there is a bug inside the PCM player and I must test a very basic fixed time scheduling on a minimized copy of your code (del del del...) to learn problems & solutions.

**Comment by JoJoBond** [comment]
You could provide float32 directly from server side and skip the whole conversion on each client. Would save you quite some trouble with the conversion.

All the player dues is schedule buffers for playback and save the ending time of the last buffer. The playback start time of each buffer is the ending time of the previous. That's pretty much it. You'll need to discard the buffers after playback, else they will reside in memory.

**Comment by ServeurpersoCom** [comment]
Using float32 on the network is way way too dirty for me lol because I love integer arithmetic, the robot is full integer maths and I do 4K lidar measure by seconds mapping on a 80MHz PIC32 and I hallucinate to see how HTML5 API are not optimized and how hard it is to do something basic like low latency audio streaming I do this in five minutes in C lol

it's been years that we should have APIs that can read all the streaming formats in 1 line of js code with all possible configuration options as in a VLC player.

**Comment by ServeurpersoCom** [comment]
OK AudioContext is a powerfull API to work with JavaScript problem, time precision and float buffering.

The code without VariSpeed is very easy to understand and can't recover from an under-run condition.
The UnderrunCallback solution is not as bad as I think but for sure we can do better

**Comment by ServeurpersoCom** [comment]
var arrayFloat32 = Float32Array.from(arrayInt16, n => n / 32768);
Give a faster conversion:) Because TypedArray are created for high performance WebGL app there is a lot of optimization (work on Android chrome)

**Comment by JoJoBond** [comment]
That's a neat method. Thanks for sharing.

**Comment by ServeurpersoCom** [comment]
If we make a voluntary error inside the mapping function "n => n / 32768" the browser speak about "native code" not js 👍

**Comment by ServeurpersoCom** [comment]
I found the bug : You must CLAMP this._NextTime >= this._SoundContext.currentTime IN ALL CASE.
Now the recovery time is instantaneous ! Sound engine can be never killed !! I don't need the external UnderrunCallback() trick anymore !

Personally I removed all VariSpeed code to keep minimalist, and I use only this :

```
    // Check if we ran out of time
    if (this._NextTime < this._SoundContext.currentTime)
        this._NextTime = this._SoundContext.currentTime;
```

Now I can keep the very low "_StartOffset = 0.1;" or lower, I remove few more code for the realtime PCM player for my robots and I past here.

Next I can add another external minimalist buffering system to add the MP3 decoder.
Tested on firefox chrome PC and Android.

Unmute function is useless now I removed it (because Safari iPhone never work even with it)

**Comment by ServeurpersoCom** [comment]
Don't need to test for this._NextTime == 0 each time because we can apply this._StartOffset in case of underrun only.

Streaming don't need a clean startup but it's impossible to ear the difference.
Result is a very smaller and very reliable code.

```
 if(this._NextTime < this._SoundContext.currentTime)
  this._NextTime = this._SoundContext.currentTime + this._StartOffset;
```

**Comment by JoJoBond** [comment]
You'll still have to check if you should play a buffer at all. Here is an example:
Lets say your JS processing freezes for 2 seconds. In the meantime buffers will still arrive via socket. Now, once your JS processing starts running again, it will playback 2 second old buffers. This will also increase the delay by 2 seconds. It would be smarter to check if your current buffer is actually so far behind that playback would not push the _NextTime into the future.

In other words each buffer we look at must be added to the _NextTime, regardless of if it is played or not.
If this._NextTime is in the future (this._NextTime > this._SoundContext.currentTime), a buffer should be schedule for playback

**Comment by ServeurpersoCom** [comment]
This code self protect from this case

```
PCMAudioPlayer.prototype.PushBuffer = function(audioBuffer) {
 var sourceNode = this._soundContext.createBufferSource();

 sourceNode.onended = function() {
  sourceNode.disconnect();
  sourceNode = null;
 };

 sourceNode.buffer = audioBuffer;

 sourceNode.connect(this._soundContext.destination);

 if(this._nextTime < this._soundContext.currentTime)
  this._nextTime = this._soundContext.currentTime + this._underrunMarginTime;
 else {
  sourceNode.start(this._nextTime);
  this._nextTime += audioBuffer.duration;
 }
};
```

**Comment by ServeurpersoCom** [comment]
Because it's the same

```
 if(this._nextTime < this._soundContext.currentTime)
  this._nextTime = this._soundContext.currentTime + this._underrunMarginTime;
 else if(this._nextTime >= this._soundContext.currentTime) { // Useless test
  sourceNode.start(this._nextTime);
  this._nextTime += audioBuffer.duration;
 }
```

**Comment by ServeurpersoCom** [comment]
My new complete audio_player.js
It can't do more than 0.4 latency, nominal = 0.2.
he starts silently, call underrun and play, call overrun and don't play "the stack of shit" and respect what you told me.

```
function PCMAudioPlayer() {
 this._soundContext = new AudioContext();
 this._nextTime = 0;
}

PCMAudioPlayer.prototype._nominalTime = 0.2;
PCMAudioPlayer.prototype._maxTime = 0.4;

PCMAudioPlayer.prototype.DecodeAudioData = function(audioData, successCallback, errorCallback) {
 return this._soundContext.decodeAudioData(audioData, successCallback, errorCallback);
};

PCMAudioPlayer.prototype.CreateBuffer = function(numberOfChannels, length, sampleRate) {
 return this._soundContext.createBuffer(numberOfChannels, length, sampleRate);
};

PCMAudioPlayer.prototype.PushBuffer = function(audioBuffer) {
 var sourceNode = this._soundContext.createBufferSource();

 sourceNode.connect(this._soundContext.destination);

 sourceNode.onended = function() {
  sourceNode.disconnect();
  sourceNode = null;
 };

 sourceNode.buffer = audioBuffer;

 if(this._nextTime == 0) {

  this._nextTime = this._soundContext.currentTime + this._nominalTime;
  sourceNode.start(this._nextTime);

 } else if(this._nextTime < this._soundContext.currentTime) {

  this.underrunCallback();
  this._nextTime = this._soundContext.currentTime + this._nominalTime;
  sourceNode.start(this._nextTime);

 } else if(this._nextTime > this._soundContext.currentTime + this._maxTime) {

  this.overrunCallback();
  this._nextTime = this._soundContext.currentTime + this._nominalTime;

 } else
  sourceNode.start(this._nextTime);

 this._nextTime += audioBuffer.duration;
};
```

**Comment by JoJoBond** [comment]
Thank you for sharing, I'm quite at the moment, I hope I will find some time tomorrow to look over it.

**Comment by Serveurperso** [comment]
There is a bug in this version : An overrun self-adjust the time and the "stack of buffers" are played the next time.
I solved this (and removed the == 0 case useless) :
I look for better

```
function PCMAudioPlayer() {
 this._soundContext = new AudioContext();
 this._nextTime = 0;
}

PCMAudioPlayer.prototype._nominalTime = 0.2;
PCMAudioPlayer.prototype._maxTime = 0.4;

PCMAudioPlayer.prototype.DecodeAudioData = function(audioData, successCallback, errorCallback) {
 return this._soundContext.decodeAudioData(audioData, successCallback, errorCallback);
};

PCMAudioPlayer.prototype.CreateBuffer = function(numberOfChannels, length, sampleRate) {
 return this._soundContext.createBuffer(numberOfChannels, length, sampleRate);
};

PCMAudioPlayer.prototype.PushBuffer = function(audioBuffer) {
 var sourceNode = this._soundContext.createBufferSource();

 sourceNode.connect(this._soundContext.destination);

 sourceNode.onended = function() {
  sourceNode.disconnect();
  sourceNode = null;
 };

 sourceNode.buffer = audioBuffer;

 if(this._nextTime < this._soundContext.currentTime) {

  this.underrunCallback();
  this._nextTime = this._soundContext.currentTime + this._nominalTime;
  sourceNode.start(this._nextTime);
  this._nextTime += audioBuffer.duration;

 } else if(this._nextTime > this._soundContext.currentTime + this._maxTime) {

  this.overrunCallback();
  sourceNode.disconnect();
  sourceNode = null;

 } else {

  sourceNode.start(this._nextTime);
  this._nextTime += audioBuffer.duration;

 }

};


```

**Comment by ServeurpersoCom** [comment]
I got it !!!! This is the perfect realtime PCM player for me !!!!! it's easy to understand and I can not do more optimized, tested again network (and js) problems, now we don't get the big bullshit of buffers when the network recover (first return case). You must add these two security on your code because at this time you have zero of them on your master branch.

- 0 is the min time, natural time under we get an Underrun condition, triggered once when there is a network shutdown. this is the first easy case I solved here because your master code go to a "tick trik" state and never recover itself.

- nominal is the normal player time, I choose middle between 0 and maxTime, the margin or originally your "StartOffset".

- and the max time (reject time) where buffers, socket.io (or any websocket/tcp!) packeds MUST be discarded. Triggered a lot after underrun recovery. This is the first case at beginning of my code. NO more useless calculation required in this critical overflow condition, I get a instantaneous recovery time on smartphones! 

you can test latency here https://www.serveurperso.com/?page=robot with the robot named "Grobot" (source is analog UHF audio with a normal power amplifier TX noise) or "Ptibot" (full digital source). you can move the servomotors to check the feedback noise without visible latency.

My audio timing setup match the raw H.264 video performance with Broadway. you can't get a better latency for piloting robots with only a web browsers on a recent smartphone and 3G+/4G or a PC with a gaming quality internet connexion.

Here is my complete tiny bulletproof audio_player.js
```
function PCMAudioPlayer() {
 this._soundContext = new AudioContext();
 this._nextTime = 0;
}

PCMAudioPlayer.prototype._nominalTime = 0.2;
PCMAudioPlayer.prototype._maxTime = 0.4;

PCMAudioPlayer.prototype.CreateBuffer = function(numberOfChannels, length, sampleRate) {
 return this._soundContext.createBuffer(numberOfChannels, length, sampleRate);
};

PCMAudioPlayer.prototype.PushBuffer = function(audioBuffer) {
 if(this._nextTime > this._soundContext.currentTime + this._maxTime)
  return;
 else if(this._nextTime < this._soundContext.currentTime) {
  this.underrunCallback();
  this._nextTime = this._soundContext.currentTime + this._nominalTime;
 }

 var sourceNode = this._soundContext.createBufferSource();

 sourceNode.connect(this._soundContext.destination);

 sourceNode.onended = function() {
  sourceNode.disconnect();
  sourceNode = null;
 };

 sourceNode.buffer = audioBuffer;
 sourceNode.start(this._nextTime);
 this._nextTime += audioBuffer.duration;
};
```
Next I can expand this code with mp3 decoder (I removed the decode wrapper temporally here)

**Comment by ServeurpersoCom** [comment]
Thank you for quoting me in your commit:) I must check your version and also make tests with MP3 (but with socket.io)

**Comment by JoJoBond** [comment]
Credit where credit is due.
